c7n/resources/appmesh.py

```diff
@@ -27,23 +27,54 @@ class AppmeshMesh(QueryResourceManager):
 
     # interior class that defines the aws metadata for resource
     class resource_type(TypeInfo):
+        # service: is used by the boto client to look up the correct API for this resource.
         service = 'appmesh'
 
         # https://docs.aws.amazon.com/service-authorization/latest/reference/list_awsappmesh.html#awsappmesh-resources-for-iam-policies   # noqa
+        # TODO: IS THIS EVEN USED AT RUNTIME?
+        # Junk value doesn't seem to affect the functionality of the extension
+        # but does fail the test PolicyMetaLint.test_valid_arn_type
+        # but that isn't a functional test.
         arn_type = "mesh"
 
         # https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-appmesh-virtualnode.html  # noqa
-        cfn_type = config_type = 'AWS::AppMesh::Mesh'
 
-        # Field in response containing the identifier used in API's.
-        # Therefore, this "id" field might be the arn field for some API's but
+        # TODO: IS THIS EVEN USED AT RUNTIME?
+        # Junk value doesn't seem to affect the functionality of the extension
+        # but does fail the test PolicyMetaLint.test_cfn_resource_validity
+        # but that isn't a functional test.
+        cfn_type = 'AWS::AppMesh::Mesh'
+
+        # TODO: IS THIS EVEN USED AT RUNTIME?
+        # Junk value doesn't seem to affect the functionality of the extension
+        # but does fail the test PolicyMetaLint.test_cfn_resource_validity
+        # but that isn't a functional test.
+        config_type = 'AWS::AppMesh::Mesh'

```

Comment:
These types are used in a few places: (config compliance filters, config execution modes, the awscc provider...)

If there is Config/CloudFormation support for a resource type, these values are _usually_ the same (which is why you often see patterns like `cfn_type = config_type = 'AWS::AppMesh::Mesh'`. There are exceptions like ACM certs, so it's still worth having separate types.

Custodian relies on shape data from the SDK and periodically updated value lists to validate these type properties, rather than adding functional tests for every resource type.

Reply:
Yeah the PolicyMetaLint does check that the config_type value is know too AWS.

The existing doc comments say these are optional but actually only 'config_type' is mandatory, whereas cfn_type is optional.

If cfn_type is optional then under what circumstances must it be provided?
Thanks

Reply:
If [CloudFormation supports a resource](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-template-resource-type-ref.html), we should record that in `cfn_type`. If that resource has explicit [Config support](https://docs.aws.amazon.com/config/latest/developerguide/resource-config-reference.html), we should record that in `config_type`.

Custodian's Config filters/modes look for a `config_type` first, and can fall back to `cfn_type`.

`cfn_type` pops up in a couple extra spots though. A non-exhaustive list there would be the awscc provider and `iam-analyzer` filter.

I wouldn't expect the `TypeInfo` comments to always contains a complete list of where those properties are used, though examples are always neat. It's pretty common that Config or CloudFormation add support well after Custodian supports a resource type, so adding `config_type` or `cfn_type` down the road is fine too.`

Reply:
Understood.